File: Omnisystem/OmniHarness/orchestrator/omniharness/substrate/governance.py
"""
Governance — the safety and control spine of the substrate.

Every autonomous run (swarm, ensemble, evolution) executes inside a Governor that
enforces resource budgets, capability policy, a tamper-evident audit log, and a
kill switch. Nothing runs unbounded; everything is recorded; the user can stop it.
"""
from __future__ import annotations

import hashlib
import json
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional


class BudgetExceeded(RuntimeError):
    """Raised when a run exceeds its resource budget."""


class PolicyViolation(RuntimeError):
    """Raised when a run attempts a disallowed model/tool/capability."""


class Aborted(RuntimeError):
    """Raised when the kill switch is tripped."""


@dataclass
class Budget:
    """Hard limits for a single governed run. 0 / None means unlimited."""
    max_model_calls: int = 200
    max_tokens: int = 2_000_000
    max_cost_usd: float = 10.0
    max_steps: int = 500
    max_wallclock_s: float = 1800.0
    max_parallel: int = 16


@dataclass
class Usage:
    model_calls: int = 0
    tokens: int = 0
    cost_usd: float = 0.0
    steps: int = 0
    started_at: float = field(default_factory=time.monotonic)

    @property
    def elapsed_s(self) -> float:
        return time.monotonic() - self.started_at

    def snapshot(self) -> Dict[str, Any]:
        return {
            "model_calls": self.model_calls,
            "tokens": self.tokens,
            "cost_usd": round(self.cost_usd, 4),
            "steps": self.steps,
            "elapsed_s": round(self.elapsed_s, 2),
        }


@dataclass
class CapabilityPolicy:
    """What a governed run is permitted to touch. Empty allow-list = allow all."""
    allowed_models: List[str] = field(default_factory=list)     # empty = any
    allowed_tools: List[str] = field(default_factory=list)      # empty = any
    denied_tools: List[str] = field(default_factory=list)
    allow_network: bool = True
    allow_filesystem_write: bool = True
    max_agents: int = 64

    def model_allowed(self, model: str) -> bool:
        return not self.allowed_models or model in self.allowed_models

    def tool_allowed(self, tool: str) -> bool:
        if tool in self.denied_tools:
            return False
        return not self.allowed_tools or tool in self.allowed_tools


class AuditLog:
    """Append-only, SHA-256 hash-chained event log — tamper-evident by construction."""

    def __init__(self) -> None:
        self._events: List[Dict[str, Any]] = []
        self._head = "0" * 64

    def append(self, kind: str, payload: Optional[Dict[str, Any]] = None) -> str:
        ts = time.time()
        body = json.dumps(payload or {}, sort_keys=True, default=str)
        digest = hashlib.sha256(f"{self._head}{ts}{kind}{body}".encode()).hexdigest()
        self._events.append({
            "seq": len(self._events),
            "timestamp": ts,
            "kind": kind,
            "payload": payload or {},
            "prev": self._head,
            "hash": digest,
        })
        self._head = digest
        return digest

    def verify(self) -> bool:
        head = "0" * 64
        for ev in self._events:
            body = json.dumps(ev["payload"], sort_keys=True, default=str)
            digest = hashlib.sha256(f"{head}{ev['timestamp']}{ev['kind']}{body}".encode()).hexdigest()
            if digest != ev["hash"] or ev["prev"] != head:
                return False
            head = digest
        return True

    def events(self) -> List[Dict[str, Any]]:
        return list(self._events)


class KillSwitch:
    """A cooperative stop signal checked by every governed loop."""

    def __init__(self) -> None:
        self._tripped = False
        self._reason = ""

    def trip(self, reason: str = "user requested stop") -> None:
        self._tripped = True
        self._reason = reason

    @property
    def tripped(self) -> bool:
        return self._tripped

    @property
    def reason(self) -> str:
        return self._reason


class Governor:
    """
    Wraps a run with budget, policy, audit, and kill-switch enforcement.
    Call `checkpoint()` at each step and `record_call(...)` after each model call.
    """

    def __init__(
        self,
        budget: Optional[Budget] = None,
        policy: Optional[CapabilityPolicy] = None,
        kill_switch: Optional[KillSwitch] = None,
        audit: Optional[AuditLog] = None,
    ) -> None:
        self.budget = budget or Budget()
        self.policy = policy or CapabilityPolicy()
        self.kill = kill_switch or KillSwitch()
        self.audit = audit or AuditLog()
        self.usage = Usage()

    # Rough cost table (USD per 1K tokens, blended) for accounting only.
    _COST_PER_1K = 0.005

    def checkpoint(self, note: str = "") -> None:
        """Enforce kill switch, wall-clock, and step limits. Call once per step."""
        if self.kill.tripped:
            self.audit.append("aborted", {"reason": self.kill.reason, "note": note})
            raise Aborted(self.kill.reason)
        self.usage.steps += 1
        if self.budget.max_steps and self.usage.steps > self.budget.max_steps:
            self.audit.append("budget_exceeded", {"limit": "max_steps", "value": self.usage.steps})
            raise BudgetExceeded(f"max_steps ({self.budget.max_steps}) exceeded")
        if self.budget.max_wallclock_s and self.usage.elapsed_s > self.budget.max_wallclock_s:
            self.audit.append("budget_exceeded", {"limit": "max_wallclock_s", "value": self.usage.elapsed_s})
            raise BudgetExceeded(f"max_wallclock_s ({self.budget.max_wallclock_s}) exceeded")

    def check_model(self, model: str) -> None:
        if not self.policy.model_allowed(model):
            self.audit.append("policy_violation", {"model": model})
            raise PolicyViolation(f"model '{model}' is not permitted by policy")
        if self.budget.max_model_calls and self.usage.model_calls >= self.budget.max_model_calls:
            self.audit.append("budget_exceeded", {"limit": "max_model_calls", "value": self.usage.model_calls})
            raise BudgetExceeded(f"max_model_calls ({self.budget.max_model_calls}) exceeded")

    def check_tool(self, tool: str) -> None:
        if not self.policy.tool_allowed(tool):
            self.audit.append("policy_violation", {"tool": tool})
            raise PolicyViolation(f"tool '{tool}' is not permitted by policy")

    def record_call(self, model: str, tokens: int) -> None:
        self.usage.model_calls += 1
        self.usage.tokens += max(0, tokens)
        self.usage.cost_usd += (max(0, tokens) / 1000.0) * self._COST_PER_1K
        self.audit.append("model_call", {"model": model, "tokens": tokens, "usage": self.usage.snapshot()})
        if self.budget.max_tokens and self.usage.tokens > self.budget.max_tokens:
            raise BudgetExceeded(f"max_tokens ({self.budget.max_tokens}) exceeded")
        if self.budget.max_cost_usd and self.usage.cost_usd > self.budget.max_cost_usd:
            raise BudgetExceeded(f"max_cost_usd ({self.budget.max_cost_usd}) exceeded")

    def parallelism(self, requested: int) -> int:
        """Clamp a requested fan-out to the policy/budget limits."""
        return max(1, min(requested, self.budget.max_parallel, self.policy.max_agents))

    def report(self) -> Dict[str, Any]:
        return {
            "usage": self.usage.snapshot(),
            "budget": self.budget.__dict__,
            "audit_valid": self.audit.verify(),
            "audit_events": len(self.audit.events()),
            "killed": self.kill.tripped,
        }

File: Omnisystem/OmniHarness/clj-orchestrator/src/omniharness/governance.clj
(ns omniharness.governance
  "Substrate governance — resource budgets, capability policy, a hash-chained
   audit log, and a kill switch. Pure, data-oriented, and thread-safe via atoms."
  (:require [clojure.string :as str])
  (:import [java.security MessageDigest]))

;; ── Audit hash chain ─────────────────────────────────────────────────────────

(defn- sha256 [^String s]
  (let [d (.digest (MessageDigest/getInstance "SHA-256") (.getBytes s "UTF-8"))]
    (apply str (map #(format "%02x" (bit-and % 0xff)) d))))

(def ^:private zero-hash (apply str (repeat 64 "0")))

(defn new-audit [] (atom {:events [] :head zero-hash}))

(defn audit-append!
  "Append a tamper-evident event; returns the new head hash."
  [audit kind payload]
  (let [ts     (System/currentTimeMillis)
        head   (:head @audit)
        body   (pr-str payload)
        digest (sha256 (str head ts kind body))]
    (swap! audit (fn [a]
                   (-> a
                       (update :events conj {:seq (count (:events a)) :ts ts
                                             :kind kind :payload payload
                                             :prev head :hash digest})
                       (assoc :head digest))))
    digest))

(defn audit-verify?
  "Replay the chain; true iff no event was mutated or reordered."
  [audit]
  (loop [head zero-hash, evs (:events @audit)]
    (if-let [ev (first evs)]
      (let [digest (sha256 (str head (:ts ev) (:kind ev) (pr-str (:payload ev))))]
        (if (and (= digest (:hash ev)) (= (:prev ev) head))
          (recur digest (rest evs))
          false))
      true)))

;; ── Budgets, policy, kill switch, governor ───────────────────────────────────

(def default-budget
  {:max-model-calls 200 :max-tokens 2000000 :max-cost-usd 10.0
   :max-steps 500 :max-wallclock-ms 1800000 :max-parallel 16})

(def default-policy
  {:allowed-models #{} :allowed-tools #{} :denied-tools #{}
   :allow-network true :allow-fs-write true :max-agents 64})

(def ^:private cost-per-1k 0.005)

(defn new-governor
  ([] (new-governor default-budget default-policy))
  ([budget policy]
   (let [g {:budget budget :policy policy
            :usage  {:model-calls 0 :tokens 0 :cost-usd 0.0 :steps 0
                     :started-ms (System/currentTimeMillis)}
            :kill   {:tripped false :reason ""}
            :audit  (new-audit)}]
     (audit-append! (:audit g) "run_start" {})
     (atom g))))

(defn trip! [gov reason]
  (swap! gov assoc :kill {:tripped true :reason reason}))

(defn checkpoint!
  "Enforce kill switch, step, and wall-clock limits. Throws ex-info on violation."
  [gov note]
  (let [{:keys [kill budget usage audit]} @gov]
    (when (:tripped kill)
      (audit-append! audit "aborted" {:reason (:reason kill)})
      (throw (ex-info (:reason kill) {:type :aborted})))
    (swap! gov update-in [:usage :steps] inc)
    (let [steps (get-in @gov [:usage :steps])
          elapsed (- (System/currentTimeMillis) (:started-ms usage))]
      (when (and (pos? (:max-steps budget)) (> steps (:max-steps budget)))
        (throw (ex-info (str "max-steps exceeded at " note) {:type :budget})))
      (when (and (pos? (:max-wallclock-ms budget)) (> elapsed (:max-wallclock-ms budget)))
        (throw (ex-info "max-wallclock-ms exceeded" {:type :budget}))))))

(defn check-model! [gov model]
  (let [{:keys [policy budget usage audit]} @gov
        allowed (:allowed-models policy)]
    (when (and (seq allowed) (not (contains? allowed model)))
      (audit-append! audit "policy_violation" {:model model})
      (throw (ex-info (str "model denied: " model) {:type :policy})))
    (when (and (pos? (:max-model-calls budget))
               (>= (:model-calls usage) (:max-model-calls budget)))
      (throw (ex-info "max-model-calls exceeded" {:type :budget})))))

(defn check-tool! [gov tool]
  (let [{:keys [policy audit]} @gov]
    (when (or (contains? (:denied-tools policy) tool)
              (and (seq (:allowed-tools policy))
                   (not (contains? (:allowed-tools policy) tool))))
      (audit-append! audit "policy_violation" {:tool tool})
      (throw (ex-info (str "tool denied: " tool) {:type :policy})))))

(defn record-call! [gov model tokens]
  (swap! gov (fn [g]
               (-> g
                   (update-in [:usage :model-calls] inc)
                   (update-in [:usage :tokens] + tokens)
                   (update-in [:usage :cost-usd] + (* (/ tokens 1000.0) cost-per-1k)))))
  (audit-append! (:audit @gov) "model_call" {:model model :tokens tokens})
  (let [{:keys [budget usage]} @gov]
    (when (and (pos? (:max-tokens budget)) (> (:tokens usage) (:max-tokens budget)))
      (throw (ex-info "max-tokens exceeded" {:type :budget})))
    (when (and (pos? (:max-cost-usd budget)) (> (:cost-usd usage) (:max-cost-usd budget)))
      (throw (ex-info "max-cost-usd exceeded" {:type :budget})))))

(defn parallelism [gov requested]
  (let [{:keys [budget policy]} @gov]
    (max 1 (min requested (:max-parallel budget)
                (if (pos? (:max-agents policy)) (:max-agents policy) requested)))))

(defn report [gov]
  (let [{:keys [usage budget audit kill]} @gov]
    {:usage usage :budget budget
     :audit-valid (audit-verify? audit)
     :audit-events (count (:events @audit))
     :killed (:tripped kill)}))
